File: modules/game.js
import { GameMap, TileType } from './map.js';
import { Bomb } from './bomb.js';
import { ExplosionEffect, BurnedWallEffect } from './effects.js';
import {
  UserMoveMessage,
  UserPlaceBombMessage,
  ServerMessageType
} from './synchronizer.js';
import { GameState } from './user.js';
import { Player } from './player.js';


function time_now() {
  return Date.now() / 1000;
}

class Game {
  constructor(players, sync, user) {
    this.map = new GameMap();
    this.players = players;
    this.sync = sync;
    this.user = user;
    this.players.forEach((player) => {
      player.game = this;
    });
    this.bombs = new Array();
    this.effects = new Array();
    this.time_left = 90;
    this.start_time = time_now();
  }

  drawScore(screen) {
    let dx = screen.sprite_size * this.map.size.x + 20;
    let dy = 30;
    let offset = 70;
    this.players.forEach((player) => {
      let color = 'black'
      switch(player.color) {
        case 0:
          color = 'aqua';
          break;
        case 1:
          color = 'lime';
          break;
        case 2:
          color = 'deeppink';
          break;
        case 3:
          color = 'orange';
          break;
      }
      screen.drawRectangle('black', dx-8, dy-18, 172, 62);
      screen.drawRectangle(color, dx-10, dy-20, 170, 60);
      screen.drawText('16px sans', 'white', dx, dy+5, player.nickname);
      screen.drawText('16px sans', 'white', dx, dy+25,
        'score: ' + player.stats);
      if (player.isDead)
        screen.drawText('40px serif', 'red', dx+10, dy+25, 'DEAD');
      dy = dy + offset;
    });
  }

  drawTime(screen) {
    let seconds_left = this.time_left - Math.floor(time_now() - this.start_time);
    console.log(seconds_left);
    screen.drawText('28px sans', 'red',
      screen.width/2 - 20, screen.height - 8, seconds_left);
  }

  render(screen) {
    this.map.draw(screen);
    this.bombs.forEach((bomb) => {
      bomb.draw(screen);
    });
    this.players.forEach((player) => {
      player.draw(screen);
    });
    this.effects.forEach((effect) => {
      effect.draw(screen);
    });
    this.drawScore(screen);
    this.drawTime(screen);
  }

  movePlayer(player, x, y, direction) {
    // TODO change hardcoded 32 with tile size
    // collison with border blocks
    if (x < 32 || x > (this.map.size.x-2)*32
        || y < 32 || y > (this.map.size.y-2)*32) {
      return false;
    }
    // collision with inner blocks
    if ((x-32)%64 < 32+32 &&
        (x-32)%64+32 > 32 &&
        (y-32)%64 < 32+32 &&
        (y-32)%64+32 > 32) return false;
    // collision with walls
    // TODO FIX THIS
    let tile_x = Math.ceil(x / 32);
    let tile_y = Math.ceil(y / 32);
    let tile = this.map.tiles.get(tile_y*this.map.size.x+tile_x);
    if (tile !== undefined && tile.type == TileType.Wall) return false;
    // send message to server
    let msg = new UserMoveMessage(direction);
    console.log('movePlayer: sendMessage');
    this.sync.sendMessage(msg);
    return true;
  }

  // NOTES: addBomb and explosion, check how bombs look like
  // and implement server then

  addBomb(x, y) {
    // calculate coordinates in grid under player
    // add new Bomb to bomb array
    let tile_x = Math.floor(x / 32);
    let tile_y = Math.floor(y / 32);
    // when server will be ready move code below to updateState
    // let bomb = new Bomb(tile_x, tile_y, this);
    // this.bombs.push(bomb);
    // send message to server
    let msg = new UserPlaceBombMessage(tile_x, tile_y);
    this.sync.sendMessage(msg);
  }

  explosion(bomb, x, y) {
    // delete bomb from array
    // add explosion effect, burning walls are a part of animation
    let index = this.bombs.indexOf(bomb);
    this.bombs.splice(index, 1);
    let explosionEffect = new ExplosionEffect(x, y, this);
    this.effects.push(explosionEffect);
    // kill hit players - synchronizer will tell us to do it anyway
  }

  destroyWalls(walls) {
    // walls - array of tile keys
    for (let i=0; i < walls.length;++i) {
      let wall = this.map.tiles.get(walls[i]);
      let burnedWallEffect = new BurnedWallEffect(wall.x, wall.y, this);
      this.effects.push(burnedWallEffect);
      this.map.tiles.delete(walls[i]);
    }
  }

  effectExpired(effect) {
    let index = this.effects.indexOf(effect);
    this.effects.splice(index, 1);
  }

  updateState() {
    // recvMessage until empty
    // apply state changes
    while(this.sync.messagesInQueue()) {
      let message = this.sync.recvMessage();
      console.log('MESSAGE! TYPE: '+ message.type);
      switch (message.type) {
        case ServerMessageType.JoinInfo:
          this.user.room_id = message.room_id;
          this.user.player.id = message.player_id;
          this.user.player.x = message.player_x;
          this.user.player.y = message.player_y;
          this.user.player.stats = message.player_stats;
          this.user.player.color = message.player_color;
          this.user.player.isDead = message.is_player_dead;
          break;
        case ServerMessageType.KickOrReject:
          // TODO change user state
          this.user.state = GameState.inMainMenu;
          break;
        case ServerMessageType.NewPlayer:
          console.log('received new player id:' + message.player_id);
          console.log('users player id: ' + this.user.player.id);
          if (message.player_id == this.user.player.id) break;
          let newPlayer = new Player(
            message.player_id,
            message.player_x,
            message.player_y,
            message.player_color
          );
          newPlayer.stats = message.player_stats;
          newPlayer.isDead = (message.is_player_dead == 1);
          newPlayer.nickname = message.player_nickname;
          newPlayer.game = this;
          this.players.push(newPlayer);
          break;
        case ServerMessageType.DelPlayer:
          let index = this.players.findIndex(p => p.id == message.player_id);
          this.players.splice(index, 1);
          break;
        case ServerMessageType.UpdatePlayer:
          let player = this.players.find(p => p.id == message.player_id);
          player.id = message.player_id;
          player.x = message.player_x;
          player.y = message.player_y;
          player.stats = message.player_stats;
          player.isDead = (message.is_player_dead == 1);
          break;
        case ServerMessageType.NewBomb:
          let new_bomb = new Bomb(message.bomb_x, message.bomb_y, this);
          this.bombs.push(new_bomb);
          break;
        case ServerMessageType.ExplodeBomb:
          let bomb = this.bombs.find(
            bomb => bomb.x == message.bomb_x && bomb.y == message.bomb_y
          );
          this.explosion(bomb, message.bomb_x, message.bomb_y);
          break;
        case ServerMessageType.NewGame:
          // new map
          this.map = new GameMap();
          this.time_left = message.time_left;
          this.start_time = time_now();
          sessionStorage.setItem("round_end",
            Math.floor(this.start_time + this.time_left));
          break;
        case ServerMessageType.MapUpdate:
          console.log('received map update');
          this.map.size = {x: message.tiles_w, y:message.tiles_h};
          this.map.tiles = message.tiles;
          break;
      }
    }
  }
}

export { Game };

File: server/game.py
import math
import time
import logging
from typing import List
from collections import deque

from aiohttp.web import WebSocketResponse

from .map import GameMap, Tile
from .bomb import Bomb
from .messages import ServerMessage, ServerMessageType, UserMessageType
from .player import Player

logger = logging.getLogger(__name__)


class Game:

    def __init__(self):
        self.players: List[Player] = []
        self.map: GameMap = GameMap()
        self.bombs: List[Bomb] = []
        self.received_messages: deque = deque()
        self.messages_to_send: deque = deque()
        self.player_move_size: int = 16
        self.counter: int = 0
        self.start_time: float = time.time()
        self.round_length: int = 90  # 90 seconds

    def step(self):
        for bomb in self.bombs:
            bomb.step()
        if time.time() - self.start_time > self.round_length:
            logger.info('round timeout')
            self.new_game()

    def move_player(self, player: Player, direction: int):
        x, y = player.x, player.y
        if direction == 1:
            y -= self.player_move_size
        elif direction == 2:
            x += self.player_move_size
        elif direction == 3:
            y += self.player_move_size
        elif direction == 4:
            x -= self.player_move_size
        # collision with border blocks
        if (x < 32 or x > (self.map.size[0]-2)*32
                or y < 32 or y > (self.map.size[1]-2)*32):
            return
        # collision with inner blocks
        if ((x-32) % 64 < 32+32 and (x-32) % 64+32 > 32 and
           (y-32) % 64 < 32+32 and (y-32) % 64+32 > 32):
            return
        # collision with walls
        tile_x = math.ceil(x / 32)
        tile_y = math.ceil(y / 32)
        if self.map.tiles[tile_y*self.map.size[0]+tile_x] != Tile.Empty:
            return
        # move
        player.x = x
        player.y = y

    def add_bomb(self, x: int, y: int):
        # TODO check if there is no wall
        for bomb in self.bombs:
            if bomb.x == x and bomb.y == y:
                return
        new_bomb = Bomb(x, y, self)
        self.bombs.append(new_bomb)
        msg = ServerMessage(
            type=ServerMessageType.NewBomb,
            bomb_x=x,
            bomb_y=y
        )
        self.messages_to_send.append(msg)

    def explode_bomb(self, bomb: Bomb):
        # destroy walls and kill players
        # up
        for i in range(bomb.strength+1):
            if (self.map.tiles[(bomb.y-i)*self.map.size[0]+bomb.x]
                    == Tile.Empty):
                for player in self.players:
                    if (math.ceil(player.x/32) == bomb.x and
                            math.ceil(player.y/32) == bomb.y-i):
                        player.is_dead = True
            elif (self.map.tiles[(bomb.y-i)*self.map.size[0]+bomb.x]
                  == Tile.Wall):
                self.map.tiles[(bomb.y-i)*self.map.size[0]+bomb.x] = Tile.Empty
                break
            elif (self.map.tiles[(bomb.y-i)*self.map.size[0]+bomb.x]
                  == Tile.Block):
                break
        # down
        for i in range(bomb.strength):
            if (self.map.tiles[(bomb.y+i)*self.map.size[0]+bomb.x]
                    == Tile.Empty):
                for player in self.players:
                    if (math.ceil(player.x/32) == bomb.x and
                            math.ceil(player.y/32) == bomb.y+i):
                        player.is_dead = True
            elif (self.map.tiles[(bomb.y+i)*self.map.size[0]+bomb.x]
                  == Tile.Wall):
                self.map.tiles[(bomb.y+i)*self.map.size[0]+bomb.x] = Tile.Empty
                break
            elif (self.map.tiles[(bomb.y+i)*self.map.size[0]+bomb.x]
                  == Tile.Block):
                break
        # left
        for i in range(bomb.strength+1):
            if self.map.tiles[bomb.y*self.map.size[0]+bomb.x-i] == Tile.Empty:
                for player in self.players:
                    if (math.ceil(player.x/32) == bomb.x-i and
                            math.ceil(player.y/32) == bomb.y):
                        player.is_dead = True
            elif self.map.tiles[bomb.y*self.map.size[0]+bomb.x-i] == Tile.Wall:
                self.map.tiles[bomb.y*self.map.size[0]+bomb.x-i] = Tile.Empty
                break
            elif (self.map.tiles[bomb.y*self.map.size[0]+bomb.x-i]
                  == Tile.Block):
                break
        # right
        for i in range(bomb.strength+1):
            if self.map.tiles[bomb.y*self.map.size[0]+bomb.x+i] == Tile.Empty:
                for player in self.players:
                    if (math.ceil(player.x/32) == bomb.x+i and
                            math.ceil(player.y/32) == bomb.y):
                        player.is_dead = True
            elif self.map.tiles[bomb.y*self.map.size[0]+bomb.x+i] == Tile.Wall:
                self.map.tiles[bomb.y*self.map.size[0]+bomb.x+i] = Tile.Empty
                break
            elif (self.map.tiles[bomb.y*self.map.size[0]+bomb.x+i]
                  == Tile.Block):
                break
        explode_msg = ServerMessage(
            type=ServerMessageType.ExplodeBomb,
            bomb_x=bomb.x,
            bomb_y=bomb.y
        )
        self.bombs.remove(bomb)
        self.messages_to_send.append(explode_msg)
        # check players left
        players_alive = 0
        if len(self.players) == 1 and self.players[0].is_dead:
            self.new_game()
        elif len(self.players) > 1:
            for player in self.players:
                if not player.is_dead:
                    players_alive += 1
            if players_alive == 1:
                for player in self.players:
                    if not player.is_dead:
                        player.stats += 1
                        self.new_game()
            elif players_alive == 0:
                self.new_game()

    def remove_player(self, player: Player):
        # remove player and add message
        del_player = ServerMessage(
            type=ServerMessageType.DelPlayer,
            player_id=player.id
        )
        self.players.remove(player)
        self.messages_to_send.append(del_player)

    def add_player(self, ws: WebSocketResponse, nickname: str) -> Player:
        starting_pos = [
            (32, 32),
            (32*(self.map.size[0]-2), 32),
            (32, 32*(self.map.size[1]-2)),
            (32*(self.map.size[0]-2), 32*(self.map.size[1]-2))
        ]
        px, py = starting_pos[len(self.players)]
        player = Player(self.counter, px, py, len(self.players), ws, stats=0)
        player.nickname = nickname
        if len(self.players) > 1:
            player.is_dead = True
        self.counter += 1
        new_player = ServerMessage(
            type=ServerMessageType.NewPlayer,
            player_id=player.id,
            player_x=player.x,
            player_y=player.y,
            player_stats=player.stats,
            player_color=player.color,
            is_player_dead=player.is_dead,
            player_nickname=player.nickname
        )
        logger.info(f'add new player[{player.id}] pos: {player.x} {player.y}')
        self.messages_to_send.append(new_player)
        self.players.append(player)
        return player

    def new_game(self):
        # set new starting positions
        logger.info('starting new game')
        starting_pos = [
            (32, 32),
            (32*(self.map.size[0]-2), 32),
            (32, 32*(self.map.size[1]-2)),
            (32*(self.map.size[0]-2), 32*(self.map.size[1]-2))
        ]
        for idx, player in enumerate(self.players):
            player.x, player.y = starting_pos[idx]
            player.is_dead = False
        # reset map and send newgame
        self.map = GameMap()
        new_game_msg = ServerMessage(
            type=ServerMessageType.NewGame,
            time_left=self.round_length
        )
        self.messages_to_send.append(new_game_msg)
        # reset timer
        self.start_time = time.time()
        # reset bombs
        self.bombs = []

    async def update_players(self):
        # send messages to players
        while self.messages_to_send:
            message = self.messages_to_send.popleft()
            for player in self.players:
                await player.send_message(message)
        # update player state
        for player in self.players:
            update_msg = ServerMessage(
                type=ServerMessageType.UpdatePlayer,
                player_id=player.id,
                player_x=player.x,
                player_y=player.y,
                player_stats=player.stats,
                is_player_dead=player.is_dead
            )
            for dest_player in self.players:
                await dest_player.send_message(update_msg)

    def get_inputs(self):
        # parse all received messages and apply changes
        logger.info('get_inputs')
        while self.received_messages:
            logger.info('received user input')
            msg = self.received_messages.popleft()
            if msg['type'] == UserMessageType.Move:
                logger.info(f'received move message: {msg}')
                self.move_player(msg['player'], msg['direction'])
            elif msg['type'] == UserMessageType.PlaceBomb:
                logger.info(f'received new bomb message: {msg}')
                self.add_bomb(msg['bomb_x'], msg['bomb_y'])
